### ISSUE TITLE
Command to display surface info

### ISSUE THREAD

**Issue by VReaperV** [Unvanquished/Unvanquished#3162]
I made an off-hand comment about this here: https://github.com/Unvanquished/Unvanquished/issues/3161?notification_referrer_id=NT_kwDOAKMSprQxMzA2MzU2NzY5ODoxMDY4NzE0Mg#issuecomment-2440031318

The more I think about it, the more I like this idea, it would work in the following way:
1. Add a `cg_showSurfaceInfoHighlight (on/off)` cvar and a `/cg_showSurfaceInfo` command (naming could be better). If the cvar is enabled, it would highlight the surface one is looking at.
2. When the command is used, it would either add the surface being looked at to a list of surfaces if it doesn't exist there, or removed from it otherwise.
3. The list of surfaces would be sent to the engine, with a reply corresponding of strings for each surface. Those strings could contain all sorts of text: `shader_t` data for that surface, `shaderStage_t` data, VBO index etc.
4. Each of these texts would be added to a text box centered on the surface origin.

Right now, debugging issues like the one linked requires lots of back and forth between frame and code debuggers, looking at shaders, adding if statements with breakpoints etc, in other words it's a very arduous process for something that is repeatedly needed. With a command like this, debugging such issues would be much easier, especially as the engine can add whatever is needed to the text for each surface.

**Comment by slipher** [comment]
I wonder if it would be better to implement the surface finding in cgame or renderer.

In cgame it might look like this:
- Extend CM trace APIs to be able to trace brushes/patches with 0 contentflags, so we can find detail brushes
- Extend trace APIs to report surface flags so we can check for SURF_NODRAW and filter out non-drawing surfaces (though I don't think these are too common)
- Add shader info to `cbrushside_t` (and whatever the patch equivalent is) and make this too available from a trace result
- Use a point trace to identify the first facing surface

It might be better to make a new trace API instead of adding all this stuff to the main one. That way we wouldn't have to add new stuff to `trace_t` or special cases for 0 content flags. Just the logic from CM_TraceThroughTree and CM_TraceThroughSurface for iterating the BSP would need to be duplicated. Then you could use the existing CM_TraceThroughBrush/CM_TraceThroughSurface to detect collisions, updating the shader info when something is found.

In the renderer it could look like this:
- After each stage of a (3D) surface is drawn, do a stencil test or something to see if the pixel in the center of the screen was written
- If the pixel was affected, add the stage to the list of relevant shader stages.

Advantages of the renderer approach:
- Could find weird stuff that wouldn't be found with the trace approach like autosprites. Or particle effects, which don't do anything weird but wouldn't be found by tracing.
- Could help debug renderer issues where some stages of a shader fail to draw anything like https://github.com/DaemonEngine/Daemon/issues/1246 or https://github.com/DaemonEngine/Daemon/issues/472.

Disadvantages:
- Probably doesn't work with material system
- Could be slow?

**Comment by VReaperV** [comment]
> After each stage of a (3D) surface is drawn, do a stencil test or something to see if the pixel in the center of the screen was written

I don't think this is possible. We'd have to read back the current stencil buffer each time and that would kill performance.

The reasoning for only getting text from engine is that it would make it easily extendable, without having to break compatibility. Adding data to brushes would also mean bloating them with a lot of otherwise unused data.

**Comment by slipher** [comment]
I have a somewhat working version of this at the `slipher/shadertrace/sync` Unvanquished/Daemon branches. `cg_showCrosshairShader 1` to use
